compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/body/resolve/FirDeclarationsResolveTransformer.kt

```diff
@@ -780,13 +781,25 @@ open class FirDeclarationsResolveTransformer(transformer: FirBodyResolveTransfor
                     dataFlowAnalyzer,
                 )
                 lambda.transformSingle(writer, expectedTypeRef.coneTypeSafe<ConeKotlinType>()?.toExpectedType())
-                val returnTypes = dataFlowAnalyzer.returnExpressionsOfAnonymousFunction(lambda)
-                    .mapNotNull { (it as? FirExpression)?.resultType?.coneType }
-                lambda.replaceReturnTypeRef(
-                    lambda.returnTypeRef.resolvedTypeFromPrototype(
-                        inferenceComponents.ctx.commonSuperTypeOrNull(returnTypes) ?: session.builtinTypes.unitType.type
-                    )
-                )
+
+                val returnStatements = dataFlowAnalyzer.returnExpressionsOfAnonymousFunction(lambda)
+                val lastStatement = lambda.body?.statements?.lastOrNull()
```

Comment:
Please run computation of `returnExpressionsExceptLast` only if `returnStatements.size > 1` to avoid unnecessary collection copying in most common cases

Reply:
Ah, I should have kept such computation overhead in mind. Thanks for the catch and suggestion!